mne/stats/regression.py

```diff
@@ -306,24 +309,21 @@ def solver(X, Y):
 
         # this is the magical part (thanks to Marijn van Vliet):
         # use toeplitz to construct series of diagonals
-        pred_arrays.append(linalg.toeplitz(samples, lags))
+        X[:, cum:cum + tmax_ - tmin_] = linalg.toeplitz(samples, lags)
+        cum += tmax_ - tmin_
 
-    big_arr = np.hstack(pred_arrays).T
     # find only those positions where at least one predictor isn't 0
-    has_val = np.where(np.any(big_arr, axis=0))[0]
+    has_val = np.where(np.any(X, axis=1))[0]
 
     # additionally, reject positions based on extreme steps in the data
     if reject is not None:
-        _, inds = _reject_data_segments(data, reject, flat, None,
+        _, inds = _reject_data_segments(data, reject, None, None,
                                         info, tstep)
         for t0, t1 in inds:
             has_val[t0:t1] = False
 
-    X = big_arr[:, has_val].T
-    Y = data[:, has_val]
-
     # solve linear system
-    coefs = solver(X, Y)
+    coefs = solver(X[has_val], data[:, has_val])
```

Comment:
unless you use slices you'll make copies. np.take is faster for access on
1st axis but does not limit memory.

The big plus in memory will be if you compute online X.T * X and X.T y as
suggested by @nsmith